### PR TITLE
Add ZSink.toQueue

### LINKED ISSUE #3735: Add ZSink.toQueue
There is `ZStream#into` operator that allows for writing elements of a stream into a queue.
Would be better to have this functionality as a `ZSink` instead because sinks can be composed.

This sink could also take a boolean parameter like `shutdownWhenFinished` to decide if the queue should be terminated after the stream is done.

I also feel that it's possible to remove `Take` from the queue type w/o loss of functionality.

### PR DESCRIPTION
**Pull request by luis3m**
Closes #3735

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 1]
```diff
@@ -838,4 +838,28 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
...
+  /**
+   * A sink that offers values to a given queue. Shuts down the queue if specified.
+   */
+  def toQueue[R, I](
+    queue: ZQueue[R, Nothing, Nothing, Any, Chunk[I], Any],
+    shutdown: Boolean
+  ): ZSink[R, Nothing, I, I, Unit] =
+    ZSink {
+      for {
+        state <- Ref.make[Chunk[I]](Chunk.empty).toManaged_
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala:851]: I don't understand the semantics here. Why do we need to accumulate all chunks offered to the queue and pass them on as a leftover? This uses unbounded memory and seems pretty unsafe to me.

Beyond that, I wouldn't expect a queue sink to result in any leftovers so the effort sounds a bit redundant; we could probably use `Nothing` for leftovers here and consider anything offered to the queue to be consumed.

What do you think @luis3m?

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 2]
```diff
@@ -838,4 +838,28 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
    */
   def sum[A](implicit A: Numeric[A]): ZSink[Any, Nothing, A, Nothing, A] =
     foldLeft(A.zero)(A.plus)
+
+  /**
+   * A sink that offers values to a given queue. Shuts down the queue if specified.
+   */
+  def toQueue[R, I](
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala:845]:
```suggestion
  def fromQueue[R, I](
```

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 3]
```diff
@@ -838,4 +838,28 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
    */
   def sum[A](implicit A: Numeric[A]): ZSink[Any, Nothing, A, Nothing, A] =
     foldLeft(A.zero)(A.plus)
+
+  /**
+   * A sink that offers values to a given queue. Shuts down the queue if specified.
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala:843]:
```suggestion
   * A sink that offers values to a given queue. Shuts down the queue on stream completion if specified.
```

**Comment by simpadjo** [comment]
I have a feeling that `ZSink.foreach` + something like `.mapM`  could be useful here

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 2 above)
Comment (simpadjo) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala:845]: Then tests for this functions must be under `suite("nstructors"){}`

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 1 above)
Comment (luis3m) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala:851]: @iravid Sorry, I'm not familiar with ZSink 😅

So you're saying the leftover part is unnecessary? I thought of leftovers based on the comment that ZSink are composable.

**Comment by iravid** [comment]
Hey @luis3m, need any help getting this over the finish line?

**Comment by luis3m** [comment]
@iravid I forgot about these, I can work on this over the weekend.

**Comment by iravid** [comment]
Thank you!

**Comment by jdegoes** [comment]
@luis3m It sounds like direction for this pull request is clear and just a bit more work necessary to get it in. Please re-open if you have the time to work more on the feature! 🙏

**Pull request closed without merging**